src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH = 1.5;
+    private Object[] array;
+    private int size;
+
+    public ArrayList() {
+        this.array = new Object[DEFAULT_CAPACITY];
+        this.size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        ensureCapacity(size + 1);
+        System.arraycopy(array, index, array, index + 1, size - index);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkBounds(index);
+        return (T) array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkBounds(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkBounds(index);
+        T removed = (T) array[index];
+        System.arraycopy(array, index + 1, array, index, size - index - 1);
+        array[--size] = null;
+        return removed;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            for (int i = 0; i < size; i++) {
+                if (array[i] == null) {
+                    return remove(i);
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (element.equals(array[i])) {
+                    return remove(i);
+                }
+            }
+        }
+        throw new NoSuchElementException("Not found this element: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void ensureCapacity(int minCapacity) {
+        if (minCapacity > array.length) {
+            int newCapacity = (int) (array.length * GROWTH);
+            if (newCapacity < minCapacity) {
+                newCapacity = minCapacity;
+            }
+            Object[] newData = new Object[newCapacity];
+            System.arraycopy(array, 0, newData, 0, size);
+            array = newData;
```

Comment:
```suggestion
            System.arraycopy(array, 0, newArray, 0, size);
```

use only array, but with correct indexes